### DISCUSSION TITLE
Getting expo-router to work in NX

### DISCUSSION THREAD

**Discussion by Evanion** [nrwl/nx#21847] [General]
I have been trying to get expo-router to work in a nx project.
So far I have gotten to the part where I'm getting an error relating to the context.
example repo: https://github.com/Evanion/nx-expo-router
This repo contains 2 apps:

* mobile - This is the expo-router project
* foobar - This is the standard app generated by the @nx/expo generator.

## Current state
Right now, I'm getting an error that says 
```
iOS Bundling failed 252ms (../../node_modules/expo-router/entry.js)
error: ../../node_modules/expo-router/_ctx.ios.js: /Volumes/projects/Regent/coin-pouch/node_modules/expo-router/_ctx.ios.js: Expected `fromDir` to be of type `string`, got `undefined`
```

### No `src` directory
I moved the app code out of the src folder, as the expo docs discurrages changing the folder structure, and when you generate a new project using `npx create-expo-app` it generates a codebase that doesn't use `src` folder. I would prefer to have one in the end though.

### Update index.js
I have updated the root index.js as per the [expo docs](https://docs.expo.dev/router/reference/troubleshooting/#expo_router_app_root-not-defined), thinking that being a monorepo the `expo-router` might get the wrong path.

```ts
import { registerRootComponent } from 'expo';

import { ExpoRoot } from 'expo-router';

export function App() {
  const ctx = require.context('./app');
  return <ExpoRoot context={ctx} />;
}

registerRootComponent(App);
```

### Declaring entry point
I have tried two approaches to declaring the entry point, both in the `apps/mobile/package.json#main` as per expo docs, and then by adding an `import 'expo-router/entry';` to the top of the root `index.js` file. This only changes the error message to point at the `Tutorial.js` file in the `expo-router` package. But it still complain that the `fromDir` is undefined rather than a string.

### Adding `EXPO_USE_METRO_WORKSPACE_ROOT`
Since when working with expo in a `yarn` workspace, you need to set an env variable to tell expo this, I added a `.env` to the `mobile` app, and added the variable, as well as updated the target in `project.json` to include it to the executor as an input. Though I suspect the that the `@nx/expo` package already does this.

```json
{
  "name": "mobile",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "apps/mobile",
  "projectType": "application",
  "targets": {
    "start": {
      "executor": "@nx/expo:start",
      "inputs": [{ "env": "EXPO_USE_METRO_WORKSPACE_ROOT" }]
    }
  },
  "tags": []
}
```

## Conclusion
I suspect that there are more people that would like to use `expo-router` with nx. But I have been unable to find any real discussions on the topic. So figured I would post something here, and hope that more people would be interested in helping out getting it fixed.

**Comment by FedeRBeije** [comment]
Update     react-native-svg-transformer  1.2.0 to 1.3.0

**Reply by Evanion** [reply]
Well... that was simple enough .. it worked! thank you!

**Reply by Stewartarmbrecht** [reply]
Worked for me too! Thank you!! Do you mind explaining why this fixed it and how you knew to do this?

**Comment by henriquechagasg** [comment]
Hi @Evanion have you succeeded on creating a development build for iOS ?
I tried to build the "mobile" project you provided here: https://github.com/Evanion/nx-expo-router, but I only could make it work with Expo Go. For development builds I'm getting the error: ` Cannot find native module 'ExpoSplashScreen', js engine: hermes`

**Reply by rufinoNL** [reply]
So I have succeeded in running it on iOS with Nx, but I have a different setup and including src (but I do have a different problem too).
What I see on your setup is that you've created an index.js to wire in everything, but you've also setup your app's package.json to point to the entry.js of expo-router. I wonder if something is wrong with that? I have no clue, just being a duck here.

My own setup is:
- app's package.json has "main": "expo-router/entry"
- I have no index.js, but I have a dir with src/app and in my project.json I have this:   "sourceRoot": "apps/fe/my-app/src", but that's not really needed as expo-router does support having a src folder "out of the box" -> https://docs.expo.dev/router/reference/src-directory/
- Also took this in regard: https://docs.expo.dev/guides/typescript/#path-aliases, which in my case looks like this in the root tsconfig.base.json:

{

  "compilerOptions": {
.. other stuff
    "paths": {
      "@@@/my-app/*": ["../nxrootfolder/apps/fe/my-app/src/*"],
      
And in my code I'll import my-app related components like:

import { MyComponent } from '@@@/my-app/some/path/toyourcomponent';

------------------------------

My issue is only on Android that it doesn't seem to work and getting:

[Nx] Unable to resolve with default PNPM resolver: ./node_modules/expo-router/entry
[Nx] Unable to resolve with any resolver: ./node_modules/expo-router/entry

And what I have tried there is to install the expo-router locally which fixes this, but nothing loads. Will try to copy all the node_modules locally to see what happens then. Again: not sure if this helps you think, but my 2 cents...

**Reply by fidestechnology** [reply]
I accidently added a new comment instead of a reply. Let me know if you have any questions.

**Reply by Evanion** [reply]
I got pulled in to a new assignment, so the project related to this issue have been put on the backburner for me. But I hope to get back in to it soon.

**Reply by FranciscoKloganB** [reply]
@Evanion Could you describe what steps of the approach above were preserved and which ones were removed?

**Reply by Chirag-kalsariya** [reply]
> Hi @Evanion have you succeeded on creating a development build for iOS ? I tried to build the "mobile" project you provided here: https://github.com/Evanion/nx-expo-router, but I only could make it work with Expo Go. For development builds I'm getting the error: ` Cannot find native module 'ExpoSplashScreen', js engine: hermes`

For this error add this package in your apps/project-name/package.json
```
    "expo-router": "*",
    "expo-splash-screen": "*",
    "react-native-safe-area-context": "*",
    "react-native-screens": "*"
```

**Comment by fidestechnology** [comment]
Sounds like issues we had in our team. Let me point you in the right direction: 

1. Add  ```"react-native-safe-area-context": "*",``` to the dependencies of your package.json within your expo nx app.
2. In the same file add ``` "main": "index.js",``` just one line under your version. 
3. Replace your ```App.tsx``` with a ```index.tsx``` within your expo apps src dir located at ```your-expo-app/src/app/index.tsx```. 
4. Change the content of your ```index.js```(not tsx) at ```your-expo-app/index.js``` to have **ONLY** one line: ```import 'expo-router/entry';```
5. If you do not have a ```.npmrc``` file at the root of your nx repo, yet, create it and add the line ```node-linker=hoisted```
6. reinstall and voila... 

This should do the trick :) I guess step 2 is the tricky part since it is not coresponsding with the docs which tell you to use ```expo-router/entry```. 

A basic version of the ```index.tsx``` at ```you-expo-app/src/app/``` could look something like this: 

```ts
import { Text } from 'react-native';

export default function Page() {
  return <Text>Home page</Text>;
}

```

you could add a ```_layout.tsx``` at the same level (```your-expo-app/src/app/_layout.tsx```) like your index.tsx (just adding is enough) to have a good place for central providers etc: 

```ts
import { Slot } from 'expo-router';
import { SafeAreaView } from 'react-native';

export default function Root() {
  return (
    <SafeAreaView>
      <Slot />
    </SafeAreaView>
  );
}
```

**Reply by sagi-passion** [reply]
Thanks [fidestechnology](https://github.com/fidestechnology).
With your help, I've managed to resolve most of the issues. However, I've encountered another problem that appears to be related to how the bundler is configuring the path for the web. Do you have any ideas on how to tackle this?

```
Web Bundling failed 8339ms (index.js)
error: ../../node_modules/expo-router/_ctx.web.js: /Users/xxxxx/xxxxx/node_modules/expo-router/_ctx.web.js: Expected `fromDir` to be of type `string`, got `undefined`
```

**Reply by ps-artem** [reply]
@sagi-passion I have the same lol

**Reply by Chirag-kalsariya** [reply]
> Sounds like issues we had in our team. Let me point you in the right direction:
> 
> 1. Add  `"react-native-safe-area-context": "*",` to the dependencies of your package.json within your expo nx app.
> 2. In the same file add ` "main": "index.js",` just one line under your version.
> 3. Replace your `App.tsx` with a `index.tsx` within your expo apps src dir located at `your-expo-app/src/app/index.tsx`.
> 4. Change the content of your `index.js`(not tsx) at `your-expo-app/index.js` to have **ONLY** one line: `import 'expo-router/entry';`
> 5. If you do not have a `.npmrc` file at the root of your nx repo, yet, create it and add the line `node-linker=hoisted`
> 6. reinstall and voila...
> 
> This should do the trick :) I guess step 2 is the tricky part since it is not coresponsding with the docs which tell you to use `expo-router/entry`.
> 
> A basic version of the `index.tsx` at `you-expo-app/src/app/` could look something like this:
> 
> ```ts
> import { Text } from 'react-native';
> 
> export default function Page() {
>   return <Text>Home page</Text>;
> }
> ```
> 
> you could add a `_layout.tsx` at the same level (`your-expo-app/src/app/_layout.tsx`) like your index.tsx (just adding is enough) to have a good place for central providers etc:
> 
> ```ts
> import { Slot } from 'expo-router';
> import { SafeAreaView } from 'react-native';
> 
> export default function Root() {
>   return (
>     <SafeAreaView>
>       <Slot />
>     </SafeAreaView>
>   );
> }
> ```

Woke for me, thank you

I write my full step process here [https://github.com/nrwl/nx/issues/23101#issuecomment-2147308325](https://github.com/nrwl/nx/issues/23101#issuecomment-2147308325)

**Reply by SimonStiph** [reply]
Thanks! I'll add I also needed to add:

  projectRoot: __dirname
  
  to the mergeConfig to get the router to know where the actual root was. Otherwise I just got a crash with the "No routes found"

**Comment by FranciscoKloganB** [comment]
We could use an official guide, by `nx` on how to setup file routing, or even updated the `generators` to take this into account.

**Reply by kaunglvlv** [reply]
https://github.com/kaunglvlv/nx-expo-router

I agree.  I couldn't get it to work using `@nx/expo` to generate the app but I was able to get it work using `create-expo-app`.  I'll update the `README` later but hopefully you can follow the commits.  Here's the high level:

1. Scaffold expo app using `create-expo-app`
2. Run `npx nx init`
3. Move app to its workspace
4. Install `expo-router` and dependencies
5. In expo app's `package.json` update `"main": "expo-router/entry"`
6. Follow [expo monorepo](https://docs.expo.dev/guides/monorepos) steps
7. Move `App.tsx` to `src/app/index.tsx`
8. Run `npx nx start <project_name>` (ex: `npx nx start mobile`)

> Edit: Looks like `expo-router` is working now. Followed @yehonadav [example](#discussioncomment-10080683)

**Comment by Chirag-kalsariya** [comment]
I think this is similar issue.

[https://github.com/nrwl/nx/issues/23101](https://github.com/nrwl/nx/issues/23101)


When I run android app using
npx nx run-android app-name

App is build and than it crash.

`E:\Heatfleet\client-monorepo\node_modules\@nx\expo\plugins\metro-resolver.js:33
        throw new Error(`Cannot resolve ${chalk.bold(realModuleName)}`);
              ^
Error: Cannot resolve ./node_modules/expo-router/entry
    at firstResolver (E:\Heatfleet\client-monorepo\node_modules\@nx\expo\plugins\metro-resolver.js:33:15)
    at firstResolver (E:\Heatfleet\client-monorepo\node_modules\@expo\cli\src\start\server\metro\withMetroResolvers.ts:108:16)
    at resolveRequest (E:\Heatfleet\client-monorepo\node_modules\@expo\cli\src\start\server\metro\withMetroResolvers.ts:137:16)
    at Object.resolve (E:\Heatfleet\client-monorepo\node_modules\metro-resolver\src\resolve.js:32:12)
    at ModuleResolver.resolveDependency (E:\Heatfleet\client-monorepo\node_modules\metro\src\node-haste\DependencyGraph\ModuleResolution.js:73:31)
    at DependencyGraph.resolveDependency (E:\Heatfleet\client-monorepo\node_modules\metro\src\node-haste\DependencyGraph.js:231:43)
    at E:\Heatfleet\client-monorepo\node_modules\metro\src\lib\transformHelpers.js:156:21
    at HmrServer._registerEntryPoint (E:\Heatfleet\client-monorepo\node_modules\metro\src\HmrServer.js:49:35)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 0)
Warning: command "expo run:android" exited with non-zero status code`


However, In browser there is no issue like this. It run well.

**Reply by Chirag-kalsariya** [reply]
This step solve this error for me.

[https://github.com/nrwl/nx/issues/23101#issuecomment-2147308325](https://github.com/nrwl/nx/issues/23101#issuecomment-2147308325)

**Comment by yehonadav** [comment]
i managed to get it working, 
here's the source code: https://github.com/yehonadav/nx-expo-router
hope that helps

**Reply by techpeace** [reply]
This ended up working in our project. Thanks so much for sharing!

**Reply by robin-scharf** [reply]
I used @yehonadav setup in April 2025 and it worked. Still not _fixed_ (updated) properly by the nx team. Be careful about the root package.json - I basically did a diff between my existing file and the packages in his file and then installed these manually via

`npx expo install {{package-name}}` in the nx project root to get the latest versions. Here's my final package.json (I have a mono repo with 3 web apps (website, web-app, admin-app), 1 server (node) and 1 mobile app:


```
{
  "name": "@XXXX/source",
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {},
  "private": true,
  "devDependencies": {
    "@eslint/compat": "^1.1.1",
    "@eslint/eslintrc": "^2.1.1",
    "@eslint/js": "^9.8.0",
    "@expo/cli": "~0.21.8",
    "@nestjs/schematics": "^10.0.1",
    "@nestjs/testing": "^11.0.12",
    "@nx/devkit": "20.6.4",
    "@nx/eslint": "20.6.4",
    "@nx/eslint-plugin": "20.6.4",
    "@nx/expo": "^20.6.4",
    "@nx/jest": "20.6.4",
    "@nx/js": "20.6.4",
    "@nx/nest": "^20.6.4",
    "@nx/next": "^20.6.4",
    "@nx/node": "20.6.4",
    "@nx/playwright": "20.6.4",
    "@nx/rollup": "^20.7.1",
    "@nx/web": "20.6.4",
    "@nx/webpack": "20.6.4",
    "@nx/workspace": "20.6.4",
    "@playwright/test": "^1.36.0",
    "@rollup/plugin-url": "^8.0.2",
    "@svgr/rollup": "^8.1.0",
    "@svgr/webpack": "^8.1.0",
    "@swc-node/register": "~1.9.1",
    "@swc/cli": "^0.6.0",
    "@swc/core": "~1.5.7",
    "@swc/helpers": "~0.5.11",
    "@tailwindcss/postcss": "^4.0.0",
    "@testing-library/dom": "10.4.0",
    "@testing-library/jest-native": "~5.4.3",
    "@testing-library/react": "16.1.0",
    "@testing-library/react-native": "~12.9.0",
    "@types/jest": "^29.5.12",
    "@types/node": "^20",
    "@types/react": "~18.3.12",
    "@types/react-dom": "~18.3.1",
    "@types/react-transition-group": "^4.4.12",
    "autoprefixer": "10.4.20",
    "babel-jest": "^29.7.0",
    "babel-preset-expo": "~12.0.1",
    "clsx": "^2.1.1",
    "eslint": "^9.8.0",
    "eslint-config-next": "^15.2.4",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-import": "2.31.0",
    "eslint-plugin-jsx-a11y": "6.10.1",
    "eslint-plugin-playwright": "^1.6.2",
    "eslint-plugin-react": "7.35.0",
    "eslint-plugin-react-hooks": "5.0.0",
    "jest": "^29.7.0",
    "jest-environment-node": "^29.7.0",
    "jest-expo": "~52.0.2",
    "jsonc-eslint-parser": "^2.1.0",
    "metro": "^0.81.0",
    "metro-config": "^0.81.0",
    "metro-resolver": "^0.81.0",
    "next-intl": "^4.0.2",
    "nx": "^20.6.4",
    "postcss": "8.4.38",
    "prettier": "^3.5.3",
    "prettier-plugin-tailwindcss": "^0.6.11",
    "react-test-renderer": "~18.3.1",
    "react-transition-group": "^4.4.5",
    "tailwindcss": "^4.0.0",
    "ts-jest": "^29.1.0",
    "ts-node": "10.9.1",
    "tslib": "^2.3.0",
    "typescript": "~5.7.2",
    "typescript-eslint": "^8.19.0",
    "verdaccio": "^6.0.5",
    "webpack-cli": "^5.1.4"
  },
  "dependencies": {
    "@expo/metro-config": "~0.19.4",
    "@expo/metro-runtime": "~4.0.0",
    "@fullcalendar/core": "^6.1.15",
    "@fullcalendar/daygrid": "^6.1.15",
    "@fullcalendar/interaction": "^6.1.15",
    "@fullcalendar/list": "^6.1.15",
    "@fullcalendar/react": "^6.1.15",
    "@fullcalendar/timegrid": "^6.1.15",
    "@nestjs/common": "^11.0.12",
    "@nestjs/core": "^11.0.12",
    "@nestjs/platform-express": "^11.0.12",
    "@nestjs/platform-fastify": "^11.0.12",
    "@react-jvectormap/core": "^1.0.4",
    "@react-jvectormap/world": "^1.1.2",
    "@tailwindcss/forms": "^0.5.9",
    "apexcharts": "^4.5.0",
    "axios": "^1.6.0",
    "expo": "~52.0.42",
    "expo-build-properties": "~0.13.2",
    "expo-constants": "~17.0.8",
    "expo-splash-screen": "~0.29.13",
    "expo-status-bar": "~2.0.1",
    "flatpickr": "^4.6.13",
    "i18next": "^24.2.3",
    "jsvectormap": "^1.6.0",
    "next": "^15.2.4",
    "react": "~18.3.1",
    "react-apexcharts": "^1.7.0",
    "react-dnd": "^16.0.1",
    "react-dnd-html5-backend": "^16.0.1",
    "react-dom": "~18.3.1",
    "react-dropzone": "^14.3.8",
    "react-i18next": "^15.4.1",
    "react-native": "^0.76.8",
    "react-native-svg": "~15.8.0",
    "react-native-svg-transformer": "~1.5.0",
    "react-native-web": "~0.19.13",
    "reflect-metadata": "^0.1.13",
    "rollup": "^4.39.0",
    "rxjs": "^7.8.0",
    "swiper": "^11.2.6",
    "tailwind-merge": "^3.1.0",
    "expo-font": "~13.0.4",
    "expo-linking": "~7.0.5",
    "expo-system-ui": "~4.0.9",
    "expo-web-browser": "~14.0.2",
    "expo-dev-client": "~5.0.19",
    "expo-router": "~4.0.20"
  },
  "nx": {
    "includedScripts": []
  },
  "overrides": {
    "@react-jvectormap/core": {
      "react": "^16.8.0 || ^17 || ^18 || ^19",
      "react-dom": "^16.8.0 || ^17 || ^18 || ^19"
    },
    "@react-jvectormap/world": {
      "react": "^16.8.0 || ^17 || ^18 || ^19",
      "react-dom": "^16.8.0 || ^17 || ^18 || ^19"
    }
  }
}

```
